core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java

```diff
@@ -65,25 +68,85 @@ public void onInit(List<? extends IndexCommit> commits) throws IOException {
     }
 
     @Override
-    public void onCommit(List<? extends IndexCommit> commits) throws IOException {
+    public void onCommit(List<? extends IndexCommit> originalCommits) throws IOException {
+        final List<TrackedIndexCommit> commits = wrapCommits(originalCommits);
         indexDeletionPolicy.onCommit(commits);
         setLastCommittedTranslogGeneration(commits);
     }
 
-    private void setLastCommittedTranslogGeneration(List<? extends IndexCommit> commits) throws IOException {
+    private void setLastCommittedTranslogGeneration(List<TrackedIndexCommit> commits) throws IOException {
         // when opening an existing lucene index, we currently always open the last commit.
         // we therefore use the translog gen as the one that will be required for recovery
-        final IndexCommit indexCommit = commits.get(commits.size() - 1);
-        assert indexCommit.isDeleted() == false : "last commit is deleted";
+        final TrackedIndexCommit indexCommit = commits.get(commits.size() - 1);
+        assert indexCommit.isMarkedAsDeleted() == false : "last commit is deleted";
         long minGen = Long.parseLong(indexCommit.getUserData().get(Translog.TRANSLOG_GENERATION_KEY));
         translogDeletionPolicy.setMinTranslogGenerationForRecovery(minGen);
     }
 
-    public SnapshotDeletionPolicy getIndexDeletionPolicy() {
-        return indexDeletionPolicy;
+    private List<TrackedIndexCommit> wrapCommits(List<? extends IndexCommit> commits) {
+        return commits.stream().map(TrackedIndexCommit::new).collect(Collectors.toList());
     }
 
-    public TranslogDeletionPolicy getTranslogDeletionPolicy() {
-        return translogDeletionPolicy;
+    /**
+     * Wraps a provided {@link IndexCommit} and tracks if it has been deleted by the policy.
+     * This tracks if {@link #delete()} has been called or not even the actual action can be ignored by {@link SnapshotDeletionPolicy}.
+     */
+    private static class TrackedIndexCommit extends IndexCommit {
```

Comment:
This sucks..  I want to see how big of a deal it is to keep things as they were. Indeed snapshotting a commit will keep it's translog around but I'm not sure anymore it's worth this kind of wrapping layers. Maybe we should invest in faster clean up on those snapshotted commits. I'll reach out to discuss.